test/integration/master_benchmark_test.go

```diff
@@ -0,0 +1,167 @@
+	errs := []error{}
+	b.StartTimer()
+	for n := 0; n < Iter; n++ {
+		// StartRC waits around for the replicas to get created
+		rc, err := framework.StartRC(string(data), uint(Parallelism), string(util.NewUUID()), m.RestClient)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		// StopRC waits around for the replicas to get deleted
+		err = framework.StopRC(rc, m.RestClient)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	b.StopTimer()
+	for err := range errs {
+		glog.Errorf("Unexpected error: %v", err)
+	}
+}
+
+// Benchmark pod listing by creating `Parallelism` pods, and listing them based on b.N
+func BenchmarkPodList(b *testing.B) {
+	b.StopTimer()
+	m := framework.NewMasterComponents(&framework.Config{nil, true})
+	defer m.Stop()
+
+	data, err := ioutil.ReadFile("benchmark-controller.json")
+	if err != nil {
+		glog.Fatalf("Unexpected error: %v", err)
+	}
+	framework.StartRC(string(data), uint(Parallelism), string(util.NewUUID()), m.RestClient)
+
+	glog.Infof("Starting pod listing benchmark with N=%d", b.N)
+	errs := []error{}
+
+	b.SetParallelism(Parallelism)
+	b.StartTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_, err := m.RestClient.Pods("test").List(labels.Everything())
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	})
+	b.StopTimer()
+	for err := range errs {
+		glog.Errorf("Unexpected error: %v", err)
+	}
+}
+
+// Benchmark roudtrip pod update->watch reflection by creating `Parallelism` pods, updating them from
+// an equal number of goroutines, and setting up a single reflector to watch changes.
+// TODO: write a benchmark with parallel watchers
+func BenchmarkWatchPods(b *testing.B) {
+	b.StopTimer()
+	m := framework.NewMasterComponents(&framework.Config{nil, true})
+	defer m.Stop()
+
+	data, err := ioutil.ReadFile("benchmark-controller.json")
+	if err != nil {
+		glog.Fatalf("Unexpected error: %v", err)
+		return
+	}
+
+	rc, err := framework.StartRC(string(data), uint(Parallelism), string(util.NewUUID()), m.RestClient)
+	if err != nil {
+		glog.Fatalf("Unable to start rc with manifest %v: %+v", string(data), err)
+		return
+	}
+
+	clientUtils := framework.ClientUtils{m.RestClient, m.StopCh, &sync.WaitGroup{}}
+	store := cache.StoreToPodLister{clientUtils.Watch(framework.CreateSelectorOrDie(client.PodHost+"=="), "pods", rc.Spec.Template.Namespace)}
+	// Make sure the reflector has started and observed existing pods
+	pods := []*api.Pod{}
+	for len(pods) < Parallelism {
+		pods, err = store.List(labels.Everything())
+		if err != nil {
+			glog.Fatalf("Unexpected error %+v", err)
+			return
+		}
+	}
+
+	glog.Infof("Starting watch pods benchmark with N=%d", b.N)
+	errs := []error{}
+	// Don't use states the rc will notice
+	lut := []api.PodPhase{api.PodRunning, api.PodPending}
+
+	b.StartTimer()
+	for i, n, status := 0, 0, lut[0]; n < Iter; n, i, status = n+1, i+1, lut[i%2] {
+		if err := clientUtils.UpdatePods(pods, status); err != nil {
```

Comment:
I think I got confused by the details of the clientUtils thing. I think moving it out of the general framework to some place like this test, and adding a comment here explaining that the update increments the waitgroup and the watch decrements it.